### PR TITLE
[feature](Nereids): eliminate left outer join by unique

### PR DESCRIPTION
**Pull request by keanji-x**
## Proposed changes

eliminate left outer join by unique

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by XieJiann** [comment]
run buildall

**Comment by XieJiann** [comment]
run buildall

**Comment by XieJiann** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateJoinByUnique.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+import org.apache.doris.nereids.trees.expressions.NullSafeEqual;
+import org.apache.doris.nereids.trees.plans.logical.LogicalJoin;
+
+/**
+ * Eliminate outer join.
+ */
+public class EliminateJoinByUnique extends OneRewriteRuleFactory {
+    @Override
+    public Rule build() {
+        return logicalProject(
```
Comment (xzj7019) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateJoinByUnique.java:31]: can we open the restriction for the project on join?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateJoinByUnique.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        ).then(project -> {
+            LogicalJoin<?, ?> join = project.child();
+            if (!join.left().getOutputSet().containsAll(project.getInputSlots())) {
+                return project;
+            }
+            if (join.getHashJoinConjuncts().stream().anyMatch(NullSafeEqual.class::isInstance)) {
+                // TODO: support null safe equals in fd and this
+                return project;
+            }
+            if (!project.getLogicalProperties().getFunctionalDependencies().isUnique(project.getOutputSet())) {
```
Comment (xzj7019) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateJoinByUnique.java:42]: Is it based on the fd derivation on both side unique key property?

File: fe/fe-core/src/test/java/org/apache/doris/nereids/rules/rewrite/EliminateJoinByUniqueTest.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        sql = "select t2.id2 from t1 left outer join t2 on t1.id1 = t2.id2";
+        PlanChecker.from(connectContext)
+                .analyze(sql)
+                .rewrite()
+                .matches(logicalJoin())
+                .printlnTree();
+    }
+
+    @Test
+    void testNull() throws Exception {
```
Comment (xzj7019) [inline fe/fe-core/src/test/java/org/apache/doris/nereids/rules/rewrite/EliminateJoinByUniqueTest.java:75]: can we add more cases like multijoin?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateJoinByUnique.java (hunk 1 above)
Comment (XieJiann) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateJoinByUnique.java:31]: only project can prune right output, where the join can be eliminated

File: fe/fe-core/src/test/java/org/apache/doris/nereids/rules/rewrite/EliminateJoinByUniqueTest.java (hunk 3 above)
Comment (XieJiann) [reply fe/fe-core/src/test/java/org/apache/doris/nereids/rules/rewrite/EliminateJoinByUniqueTest.java:75]: multijoin has no difference with the single join

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateJoinByUnique.java (hunk 2 above)
Comment (XieJiann) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateJoinByUnique.java:42]: yes

**Review by xzj7019 — APPROVED** [review]

**Review by jackwener — DISMISSED** [review]

**Comment by XieJiann** [comment]
run buildall

**Comment by XieJiann** [comment]
run buildall

**Comment by XieJiann** [comment]
run buildall

**Comment by XieJiann** [comment]
run buildall

**Review by jackwener — APPROVED** [review]

**Pull request merged**